pkg/odo/cli/init/init.go

```diff
@@ -0,0 +1,306 @@
+		defer downloadSpinner.End(false)
+		params := util.HTTPRequestParams{
+			URL: URL,
+		}
+		devfileData, err := util.DownloadFileInMemory(params)
+		if err != nil {
+			return err
+		}
+		err = o.fsys.WriteFile(dest, devfileData, 0644)
+		if err != nil {
+			return err
+		}
+		downloadSpinner.End(true)
+	} else {
+		downloadSpinner := log.Spinnerf("Copying devfile from %q", URL)
+		defer downloadSpinner.End(false)
+		content, err := o.fsys.ReadFile(URL)
+		if err != nil {
+			return err
+		}
+		info, err := o.fsys.Stat(URL)
+		if err != nil {
+			return err
+		}
+		err = o.fsys.WriteFile(dest, content, info.Mode().Perm())
+		if err != nil {
+			return err
+		}
+		downloadSpinner.End(true)
+	}
+
+	return nil
+}
+
+// downloadRegistry downloads a devfile from the provided registry and saves it in dest
+func (o *InitOptions) downloadRegistry(registryName string, devfile string, dest string) error {
+	downloadSpinner := log.Spinnerf("Downloading devfile %q from registry %q", devfile, registryName)
+	defer downloadSpinner.End(false)
+	registries := o.preferenceClient.RegistryList()
+	var registry preference.Registry
+	var found bool
+	for _, registry = range *registries {
+		if registry.Name == registryName {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return fmt.Errorf("unable to find the registry with name %q", registryName)
+	}
+
+	err := registryLibrary.PullStackFromRegistry(registry.URL, devfile, dest, segment.GetRegistryOptions())
+	if err != nil {
+		return err
+	}
+	downloadSpinner.End(true)
+	return nil
+}
+
+// downloadStarterProject downloads the starter project referenced in devfile and stores it in dest directory
+// WARNING: This will first remove all the content of dest.
+func (o *InitOptions) downloadStarterProject(devfile parser.DevfileObj, project string, dest string) error {
+	projects, err := devfile.Data.GetStarterProjects(common.DevfileOptions{})
+	if err != nil {
+		return err
+	}
+	var prj v1alpha2.StarterProject
+	var found bool
+	for _, prj = range projects {
+		if prj.Name == project {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return fmt.Errorf("starter project %q does not exist in devfile", project)
+	}
+	downloadSpinner := log.Spinnerf("Downloading starter project %q", prj.Name)
+	err = component.DownloadStarterProject(&prj, "", dest, false)
+	if err != nil {
+		downloadSpinner.End(false)
+		return err
+	}
+	downloadSpinner.End(true)
+	return nil
+}
+
+// NewCmdInit implements the odo command
+func NewCmdInit(name, fullName string) *cobra.Command {
+	backends := []params.ParamsBuilder{
+		&params.FlagsBuilder{},
+		params.NewInteractiveBuilder(asker.NewSurveyAsker(), catalog.NewCatalogClient()),
+	}
+	prefClient, err := preference.NewClient()
+	if err != nil {
+		odoodoutil.LogErrorAndExit(err, "unable to set preference, something is wrong with odo, kindly raise an issue at https://github.com/redhat-developer/odo/issues/new?template=Bug.md")
```

Comment:
Yes, it would be nice, but not sure about the method to do this:
- declare a constant string somewhere, and reuse it instead of the literal string?
- Write a NewClientAssert() wrapper, that will call LogErrorAndExit if error, and use NewclientAssert everywhere instead of Newclient?

Reply:
Or create a new type that implements `Error()` method? I remember @valaparthvi mentioning the same somewhere on this or other PR of yours. I like the idea of having `errors.go`. What do you think is general practice?